pkg/apis/apps/v1alpha1/imagepullnode_types.go

```diff
@@ -0,0 +1,193 @@
+}
+
+// ImageTagPullStatus defines the pulling status of an image tag
+type ImageTagPullStatus struct {
+	// Represents the image tag.
+	Tag string `json:"tag"`
+
+	// Represents the image pulling task phase.
+	Phase ImagePullPhase `json:"phase"`
+
+	// Represents the pulling progress of this tag, which is beetween 0-100. There is no guarantee
+	// of monotonic consistency, and it may be a rollback due to retry during pulling.
+	Progress int32 `json:"progress,omitempty"`
+
+	// Represents time when the pulling task was acknowledged by the image puller.
+	// It is not guaranteed to be set in happens-before order across separate operations.
+	// It is represented in RFC3339 form and is in UTC.
+	// +optional
+	StartTime *metav1.Time `json:"startTime,omitempty"`
+
+	// Represents time when the pulling task was completed. It is not guaranteed to
+	// be set in happens-before order across separate operations.
+	// It is represented in RFC3339 form and is in UTC.
+	// +optional
+	CompletionTime *metav1.Time `json:"completionTime,omitempty"`
+
+	// Represents the summary informations of this node
+	// +optional
+	Message string `json:"message,omitempty"`
+}
+
+// ImagePullPhase defines the tasks status
+type ImagePullPhase string
+
+const (
+	// ImagePhaseWaiting means the task has not started
+	ImagePhaseWaiting ImagePullPhase = "Waiting"
+	// ImagePhasePulling means the task has been started, but not completed
+	ImagePhasePulling ImagePullPhase = "Pulling"
+	// ImagePhaseSucceeded means the task has been completed
+	ImagePhaseSucceeded ImagePullPhase = "Succeeded"
+	// ImagePhaseFailed means the task has failed
+	ImagePhaseFailed ImagePullPhase = "Failed"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +genclient:nonNamespaced
+
+// ImagePullNode is the Schema for the imagepullnodes API
+// +k8s:openapi-gen=true
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="DESIRED",type="integer",JSONPath=".status.desired",description="Number of all images on this node"
+// +kubebuilder:printcolumn:name="PULLING",type="integer",JSONPath=".status.pulling",description="Number of image pull task active"
+// +kubebuilder:printcolumn:name="SUCCEED",type="integer",JSONPath=".status.succeeded",description="Number of image pull task succeeded"
+// +kubebuilder:printcolumn:name="FAILED",type="integer",JSONPath=".status.failed",description="Number of image pull tasks failed"
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp",description="CreationTimestamp is a timestamp representing the server time when this object was created. It is not guaranteed to be set in happens-before order across separate operations. Clients may not set this value. It is represented in RFC3339 form and is in UTC."
+type ImagePullNode struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ImagePullNodeSpec   `json:"spec,omitempty"`
```

Comment:
Do you need to specify the node name in the spec? or how will the node daemon find the corresponding crd?

Reply:
The CR will be automatically created by the controller, and use the node name as the CR name.